### PR TITLE
[docs] Add support for external markdown

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

This PR fixes the issue of diverging documentation between Ray Docs and ecosystem library readmes which live in separate repos (eg. xgboost_ray). This is achieved by adding an extra step before the docs build process starts that downloads the readmes of specified ecosystem libraries from their GitHub repositories. The files are then preprocessed by a very simple parser to allow for differences between GitHub and Docs markdowns.

In summary, this makes the markdown files in ecosystem library repositories single sources of truth and removes the need to manually keep the doc pages up to date, all the while allowing for differences between what's rendered on GitHub and in the Docs.

See https://github.com/ray-project/xgboost_ray/pull/204 & https://ray--23505.org.readthedocs.build/en/23505/ray-more-libs/xgboost-ray.html for an example.

Needs https://github.com/ray-project/xgboost_ray/pull/204 and https://github.com/ray-project/lightgbm_ray/pull/30 to be merged first.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/README.md [hunk 1]
```diff
@@ -81,3 +81,47 @@ Using caching, this means the first time you build the documentation, it might t
...
+
+In order to avoid a situation where duplicate documentation files live in both the docs folder
+in this repository and in external repositories of ecosystem libraries (eg. xgboost-ray), you can
+specify Markdown files that will be downloaded from other GitHub repositories during the build process.
+
+In order to do that, simply edit the `EXTERNAL_MARKDOWN_FILES` list in `source/custom_directives.py`
+using the format in the comment. Before build process, the specified files will be downloaded, preprocessed
+and saved to given paths. The build process will then proceed as normal.
+
+Because GitHub Markdown and MyST differ in syntax, simple preprocessing is performed to allow for differences
```
Comment (maxpumperla) [inline doc/README.md:95]: MyST is common markdown compliant. Might be good to mention that here. I often find it easy to restrict markdown dialect usage, it's all just syntactic sugar in the end.

File: doc/source/custom_directives.py [hunk 2]
```diff
@@ -174,3 +176,71 @@ def mock_modules():
 
     for mod_name in CHILD_MOCK_MODULES:
         sys.modules[mod_name] = ChildClassMock()
+
+
+# Add doc files from external repositories to be downloaded during build here
+# (repo, ref, path to get, path to save on disk)
+EXTERNAL_MARKDOWN_FILES = [
+    ("ray-project/xgboost_ray", "master", "README.md", "ray-more-libs/xgboost-ray.md"),
+    ("ray-project/lightgbm_ray", "master", "README.md", "ray-more-libs/lightgbm-ray.md"),
```
Comment (maxpumperla) [inline doc/source/custom_directives.py:185]: linter seems to have problems with those lines

File: doc/README.md [hunk 3]
```diff
@@ -81,3 +81,47 @@ Using caching, this means the first time you build the documentation, it might t
...
+in rendering on GitHub and in docs. You can use two commands (`$UNCOMMENT` and `$REMOVE`/`$END_REMOVE`)
+in the Markdown file, specified in the following way:
+
+### `$UNCOMMENT`
+
+GitHub:
+
+```html
+<!--$UNCOMMENTthis will be uncommented--> More text
```
Comment (maxpumperla) [inline doc/README.md:104]: I wonder if we can somehow make this convention easier in a follow-up PR. It'd be great to add more external content from trusted sources, without them having to add our special commenting syntax.

**Review by maxpumperla — APPROVED** [review]
Fantastic idea, really like the preprocessor system!

File: doc/source/ray-more-libs/lightgbm-ray.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<!--
```
Comment (amogkam) [inline doc/source/ray-more-libs/lightgbm-ray.md:1]: do we need one for xgboost as well?

**Review by amogkam — COMMENTED** [review]
@Yard1 as I'm developing docs in external repos, is there a way to see how it will be rendered on the Ray docs? If so, could we add this to the README section?

File: doc/source/ray-more-libs/lightgbm-ray.md (hunk 4 above)
Comment (Yard1) [reply doc/source/ray-more-libs/lightgbm-ray.md:1]: There is a file there for xgboost already

**Comment by Yard1** [comment]
@amogkam I'll add a convenience script to convert an external markdown file to MyST file and then you can build the docs locally with that included to see how it will look like

**Review by krfricke — COMMENTED** [review]
Great idea! And I generally like the implementation. Just two questions:

1) Instead of uncommenting e.g. headlines, can we just have a "template" on the Ray docs side that replaces some body part with the fetched content? That way we can have more content on Ray (e.g. the API docs, which are not used in the library READMEs at all but only in the Ray docs).
2) Can we just convert the READMEs with some converter instead of providing e.g. two versions of the same table?

**Comment by Yard1** [comment]
@krfricke The problem with conversion is that there are some elements that only make sense in the docs (eg. sphinx headers). As for the tables specifically, I believe there wouldn't be any issues if we left them in markdown, this is just how it is in the docs currently.

I toyed with the template idea before through the use of sphinx directives, but unfortunately it's not straightforward (especially with MyST having issues with custom Include-like directives). I think the simple uncomment/remove combo provides the greatest flexibility. Templates etc. could potentially get messy and complicated. Plus we are in control of the external docs, so we can make any changes we need anytime.

**Pull request merged**